292 - Fix: Implement the dataset loader for polling the latest dataset state for the dataset page UI

nozomione opened the pull request:
## Issue Number
**NOTE:** The **changes in 5 files** were simply the results of fixing hook imports and argument orders, deletion of a file, and state imports. 

Closing #292

## Purpose/Implementation Notes
Implemented logic for polling the latest state of the processing dataset, and based on that state, render the appropriate view (processing/ready for download/error) on the dataset page.

- Created the `useResourceLoader` hook (refactored [`useOneoffExperiment`](https://github.com/AlexsLemonade/refinebio-web/pull/215) hook) for downloading files for both user-created datasets and one-off experiments datasets. 

Please see the latest UI [here](https://refinebio-clz6yymct-ccdl.vercel.app/).

For the user-defined datasets (My Dataset):
**Step 1:** Add samples to My Dataset 
**Step 2:** Start the dataset download process by clicking the "Download" button in the dataset page
**Step 3:** Once the dataset processing is completed, the application renders the appropriate UI (success/error)

![Screenshot 2023-11-20 at 9 17 59 PM](https://github.com/AlexsLemonade/refinebio-web/assets/31800566/de185687-510d-4c3b-85be-f35d1cb59b03)

<br/><br/>
For the one-off experiment datasets:
**Step 1:** Click the "Download Now" button in a search card on the search results page to start the download process
**Step 2:** The application renders the processing budge while the one-off experiment dataset being processed

![Screenshot 2023-11-21 at 2 01 37 PM](https://github.com/AlexsLemonade/refinebio-web/assets/31800566/b95b36e2-4b49-4964-bec6-8426fc22b5e6)

(search results page: while processing)

![image](https://github.com/AlexsLemonade/refinebio-web/assets/31800566/807e9b0a-8486-42ac-b415-a6206bd6303a)

(experiment page: while processing)

![image](https://github.com/AlexsLemonade/refinebio-web/assets/31800566/6445860a-5a04-4cae-8f49-776f5c5c49df)



## Types of changes
- Breaking change (fix or feature that would cause existing functionality to not work as expected)

## Functional tests

Tested the implementation and made sure it works for both My Dataset download and the dataset download for one-off experiment. 

## Checklist

<!-- Put an `x` in the boxes that apply. -->

- [x] Lint and unit tests pass locally with my changes
- [ ] I have added tests that prove my fix is effective or that my feature works
- [ ] I have added necessary documentation (if appropriate)
- [ ] Any dependent changes have been merged and published in downstream modules

davidsmejia commented on src/components/Dataset/DatasetPageHeader.js:
```diff
@@ -32,8 +32,7 @@ const Block = ({ children }) => {
 
-export const DatasetPageHeader = ({ dataset }) => {
-  const { error } = useDatasetManager()
+export const DatasetPageHeader = ({ dataset, hasError }) => {
```
I'm trying to understand why we must pass in `hasError` and wouldn't be able to derive it from the `dataset` prop.

nozomione replied:
The `useResourceLoader` hook is used by different components (including this `DatasetPageHeader`, [DownloadNowModal](https://github.com/AlexsLemonade/refinebio-web/blob/b45ba3d342e247f094c3ece4bcb79bc509e5a437/src/components/shared/SearchCard/SearchCardAction/DownloadNowModal.js#L24-L26), and the [experiment](https://github.com/AlexsLemonade/refinebio-web/blob/b45ba3d342e247f094c3ece4bcb79bc509e5a437/src/components/shared/SearchCard/SearchCardAction/DownloadNowModal.js#L24-L27) page components) to "watch" the latest processing status of resources (a user-custom dataset and an one-of experiment download's dataset). The `hasError` state is set to true if API requests fail while polling the data, and it could be used to display a error message in UI etc.  

In terms of the `dataset` state, it's a part of `useDatasetManager` and is used to render the dataset's data in UI.

This implementation (utilizing the `hasError` state) also matches other hooks' pattern ([`useComendia`](https://github.com/AlexsLemonade/refinebio-web/blob/118c281d226a70d6cdc972c290b87b25c93c6b18/src/hooks/useCompendia.js#L20), [`useLoader`](https://github.com/AlexsLemonade/refinebio-web/blob/118c281d226a70d6cdc972c290b87b25c93c6b18/src/hooks/useLoader.js#L10), and [`useSamplesTable`](https://github.com/AlexsLemonade/refinebio-web/blob/118c281d226a70d6cdc972c290b87b25c93c6b18/src/hooks/useSamplesTableManager.js#L17), and the helper [`fetchSearch`](https://github.com/AlexsLemonade/refinebio-web/blob/118c281d226a70d6cdc972c290b87b25c93c6b18/src/helpers/fetchSearch.js#L62) for search results page that fetches API data via its `getInitialProps`). 

For this, I used the `useResourceLoader` hook's `hasError` (instead of `useDatasetManager`'s `error` object - please see the PR https://github.com/AlexsLemonade/refinebio-web/pull/297) and pass it to `DatasetPageHeader` (which is responsible for re-rendering UI based on the most recent dataset status that was last polled from API) via the [`dataset_id`](https://github.com/AlexsLemonade/refinebio-web/blob/b45ba3d342e247f094c3ece4bcb79bc509e5a437/src/pages/dataset/%5Bdataset_id%5D.js#L33-L38) page component.

davidsmejia reviewed (commented):
Left some initial thoughts here. They mostly focus on the `useResourceLoader` hook.

nozomione commented on src/components/Dataset/DatasetPageHeader.js:
```diff
@@ -46,46 +45,36 @@ export const DatasetPageHeader = ({ dataset }) => {
+                Shared Dataset
+              </Heading>
+            </Box>
+          )
```
I've reverted the JSX changes (at 706548df0f631f4b63a7708d2882de9168563564) and it now uses early returns 👍

nozomione commented on src/hooks/useResourceLoader.js:
Renamed af3b4c72d78188d80344fbfa73f991f148109e9b.

nozomione commented on src/hooks/useResourceLoader.js:
```diff
@@ -0,0 +1,80 @@
+
+  // id: a dataset ID || an experiment accession code
+  const getProcessingResource = (id) => {
+    const keyToFind = oneOffExperiment ? 'accesionCode' : 'datasetId'
```
I've remove this optional parameter `oneOffExperiment` and instead, now using the areValidAccessionCodes helper to check whether the given `resourceId` is an accession code or a dataset ID (deb7269a9f2a098b8b865c6bb3de6b9e9b55e189).

nozomione replied:
*also thank you for this great idea!

nozomione commented on src/hooks/useResourceLoader.js:
```diff
@@ -0,0 +1,80 @@
+export const useResourceLoader = (resourceId, oneOffExperiment = false) => {
+  const { getDataset } = useDatasetManager()
+  const { processingResources, setProcessingResources } = useRefinebio()
+  const [hasError, setHasError] = useState(false)
```
As mentioned in the [comment](https://github.com/AlexsLemonade/refinebio-web/pull/294#discussion_r1426015939), this implementation (which makes use of the `hasError` state) also matches the pattern of other hooks, so let's create a separate issue to refactor this hook pattern to make the error state as part of a context 👍

nozomione commented on src/hooks/useDatasetManager.js:
```diff
@@ -89,7 +89,9 @@ export const useDatasetManager = () => {
-    const isCurrentDatasetId = id === datasetId
+    const isCurrentDatasetId = id && id === datasetId
+    const datasetIdToUse =
+      !id || !isCurrentDatasetId ? await createDataset() : id
```
```javascript
/* useDatasetManager.js */
// it takes parameters for a dataset ID and download options
  const startProcessingDataset = async (id, options) => { … }
```

This method is also used for one-time experiment downloads. 

Since an one-off has no associated dataset ID by default, `null` was passed as the first argument  `id`.  For this, the `!id ||` check was also added in this PR - commit at f8291c70385c840c72e479a0b4e52cd452a2f33d.

To revert this commit change, I’ve updated the parameter order and made `id`  an optional parameter with its default value (as below), removed the extra logical OR, and adjusted argument order in the files that use this method (6f99fd837d813739aa7ff33f59a048bb2f562ede).

```javascript
/* useDatasetManager.js */
  const startProcessingDataset = async (options, id = null) => { … }
```

nozomione commented:
@davidsmejia this PR is ready for another look.

davidsmejia reviewed (commented):
Left a comment about removing the error on polling from the dataset page header. We should also look at how the `usePollDatasetStatus` hook should handle distinguishing between one off downloads and the "my dataset" dataset.

It would make sense to me to move the use of the polling hook to the `useDatasetManager` to allow for polling on that state instead of creating an interface that takes either an experiment accession or a dataset id. That way we wouldn't have to do inference on the type of ID as we should only be polling once the dataset is created.

davidsmejia commented on src/hooks/usePollDatasetStatus.js:
```diff
@@ -0,0 +1,84 @@
+    const response = await getDataset(datasetId)
+
+    setHasError(response?.ok === false)
+    setLatestDatasetState(response)
```
We should only update this on successful fetches.

nozomione commented on src/components/Dataset/DatasetPageHeader.js:
```diff
@@ -46,7 +45,7 @@ export const DatasetPageHeader = ({ dataset }) => {
   const isProcessingError = dataset?.success === false // 'success' may be null
 
-  if (isProcessingError || error) {
+  if (isProcessingError || hasError) {
```
Removed at 09fdc6e996d56ccf7c41cd47f6150221ca7eb8b0.

nozomione commented:
@davidsmejia I've applied the feedback and this PR is ready for another look!

nozomione commented:
> Let's remove the usage of "resource" in variable names and function names associated with this PR.

What term would you suggest in replace of `resource` ?

The `dataset` and `datasetId` are taken by `DatasetManager`, and we're also passing  an accession code to this hook, we want to use a name that represents both of those resources. Perhaps what do you think about `selectedDataset`?

> Break up the "getters" since we will know when using this hook if we want to check if an accession has a processing dataset or if we are checking by dataset id explicitly.

Currently the resource id check is done by using the regular expression:
https://github.com/AlexsLemonade/refinebio-web/blob/09fdc6e996d56ccf7c41cd47f6150221ca7eb8b0/src/hooks/usePollDatasetStatus.js#L49-L51

Could you elaborate on this one?

Thank you David!

davidsmejia commented:
> > Let's remove the usage of "resource" in variable names and function names associated with this PR.
> 
> What term would you suggest in place of `resource` ?
> 
> The `dataset` and `datasetId` are taken by `DatasetManager`, and we're also passing an accession code to this hook, we want to use a name that represents both of those resources. Perhaps what do you think about `selectedDataset`?
> 
> > Break up the "getters" since we will know when using this hook if we want to check if an accession has a processing dataset or if we are checking by dataset id explicitly.
> 
> Currently the resource id check is done by using the regular expression:
> 
> https://github.com/AlexsLemonade/refinebio-web/blob/09fdc6e996d56ccf7c41cd47f6150221ca7eb8b0/src/hooks/usePollDatasetStatus.js#L49-L51
> 
> Could you elaborate on this one?
> 
> Thank you David!

I think these questions are related so I am going to provide a single response to all of them first.

The `useDatasetManager` should be responsible for creating, querying, configuring, and describing a dataset. Essentially activity related to a specific dataset.

The `usePollDatasetStatus` is the collection of logic around polling a specific dataset. This currently, and correctly, uses `useDatasetManager` under the hood to perform the query.

The change I would like to see here is how we associate a specific accession code with a dataset.

Currently `usePollDatasetStatus` does the following in this PR:
- Initializes the hook with either an accession code or a dataset id.
- fetches the latest state on mount
- provides helper methods for managing a list of currently polled datasets, either by ID of accession code.
- stores an array of objects that contain datasetid and optionally accession code for association

I would like to see the ambiguous argument (dataset id / accession code) removed from the initialization of the hook and from the `add.

We would accomplish this by:
- removing the initialization argument from the `usePollDatasetStatus` hook
- remove the `useEffect` that checks if the ID can be found and refreshes if it is lines 15-19
- move `processingDatasets` (old name `processingResources`) `useDatasetManager`
- Add a local storage state in RefinebioContext / exposed in hook with value of `datasetAccessions` that will default to an empty dict. The keys will be accession codes and the values will be dataset ids.
- add new method `pollDatasetId`
- add new method `pollDatasetAccession`
  - these functions should just set the datasetid and that is what the useEffect/polling logic should check

nozomione commented:
The following updates were completed:
> - remove the `useEffect` that checks if the ID can be found and refreshes if it is lines 15-19
> - move `processingDatasets` (old name `processingResources`) `useDatasetManager`

And now that we've moved `processingDatasets` to `useDatasetManager`, I've also: 
- moved `addDatasetToProcessingDatasets`(previously `addProcessingDataset` from `usePollDatasetStatus`) to `useDatasetManager`
- call  `addDatasetToProcessingDatasets` from `startProcessingDataset` to handle the addition of a new processing dataset (no longer calling it from two separate locations)

(see commit at be12eb546c9d68058a5c506f981852100c7e0057)

For the following items, I have a few questions: 

**Q 1:** How do we determine which dataset ID to poll inside the hook if we remove the hook's parameter?
> - removing the initialization argument from the `usePollDatasetStatus` hook

<details>
   <summary><strong>Reasons for passing an ID as an argument to <code>usePollDatasetStatus</code> hook</strong></summary><br/>

- To quickly check the given ID and initiate a polling request on component mount 
- To keep the polling logic inside the hook as much as possible (including ID check, start/refresh/end polling)

```js 
usePollDatasetStatus(processingId) // either a dataset ID or an one-off experiment accession code
```

When the following UIs are viewed by the user (on component mount), this hook checks the given ID, starts polling requests, and display the processing UI while it's being processed:

- Search card and its download modal in the search results page (passing an accession code to the hook):
![Screenshot 2024-06-03 at 5 14 09 PM](https://github.com/AlexsLemonade/refinebio-web/assets/31800566/db68b1a3-42bc-4cbf-89ea-d9b008afee09)

- Experiment page for a specific accession code (passing an accession code to the hook):
![Screenshot 2024-06-03 at 5 14 20 PM](https://github.com/AlexsLemonade/refinebio-web/assets/31800566/280ba0b7-f6b3-4611-9a8e-41b799dd4c94)

- Dataset page that renders a specific dataset ID (passing a dataset ID in the URL to the hook):
(e.g., `/dataset/695fd15-2f7c-4b5c-8b39-212f9ee7b21e`)
![Screenshot 2024-06-03 at 5 15 06 PM](https://github.com/AlexsLemonade/refinebio-web/assets/31800566/a4290357-db02-438e-8a4b-632fef05c36c)

<hr/>

</details>

By removing the argument for the hook (ID to check) and adding the next two methods:
> - add new method pollDatasetId
> - add new method pollDatasetAccession
>   - these functions should just set the datasetid and that is what the useEffect/polling logic should check

**Q 2:**  Where would you like the dataset ID to be set and start the polling request? 

To put another way, if this hook no longer accepts an ID as an argument, then at first, the ID check must to be done outside of the hook (a component that uses this hook by calling these new methods, `useEffect` etc). And possibly adding a new state to the hook to store that obtained dataset ID so that we can use it inside the hook to start polling. 

I suggest preserving the hook's argument to maintain the quick ID check and the polling logic within the hook as much as possible.

I'd like to hear your input on this. Please let me know, thank you David!

davidsmejia commented:
Ok, I think your questions stem from the same point so let me try clarifying this first and if still unclear I will try to tackle the questions separately.

Add a local storage state in RefinebioContext / exposed in hook with value of datasetAccessions that will default to an empty object. The keys will be accession codes and the values will be dataset ids.

So you don't have to care that it is a valid accession code here. When you add the dataset with optional accession code, you populate a separate object that functions like a look up table. No need for the regex.


So update what you have in `useDatasetManager` starting at line 34
Here you only need to assign a flat value of processing datasets and an object to look up by accession code.
So later in the application you will always know one of two things when you need to check if a dataset is processing:
- The dataset id, this is obvious
- The accession code which can be looked up to get the dataset id
 
`useDatasetManager.js`
```javascript
 const addToProcessingDatasets = (id, accessionCode) => {
    if (accessionCode) {
      setDatasetAccessions({...datasetAccessions, [accessionCode]: id})
    }
    setProcessingDatasets((prev) => {
      if (prev.includes(id)) return prev
      return [...prev, id]
    })
  }
```

Then you want to poll for a specific dataset. This should be done by calling one of the two functions below:
`usePollDatasetStatus.js`
```javascript

// this is the state variable that will trigger a rerender when updated
const [polledDatasetId, setPolledDatasetId] = useState(null)
const [datasetState, setDatasetState] = useState(null)
const timerRef = useRef(null)


// check if the hook should poll a dataset
  useEffect(() => {
    if (polledDatasetId) {
      timerRef.current = setInterval(() => {
        refreshProcessingDataset()
      }, 1000 * 60)
    }
    return () => {
      clearInterval(timerRef.current)
    }
  }, [polledDatasetId])

// by dataset ID 
pollDatasetId(datasetID) {
  // the dataset doesnt need to be polled anymore because it has been removed
  if (processingDatasets.includes(datasetID)) {
    setPolledDataset(datasetID)
  }
}

// tell the hook to watch a specific accession
/// call the above method if there is a match
pollDatasetAccession(accession) {
  if (datasetAccessions.hasOwnProperty(accession)) {
    pollDatasetId(datasetAccessions[accession])
  }
}

```

Let me know if you can see how this would work. Each hook would only be responsible for tracking the state of a single dataset. We should have a separate cleanup function ie `removeProcessingDataset` that only takes dataset id and removes it from the list of processing datasets and the datasetAccessions object.

nozomione commented:
The following updates were completed:
> - removing the initialization argument from the usePollDatasetStatus hook
> 

Now the hook no longer takes an ID argument and instead uses a new state, `polledDatasetId`, to set the processing dataset ID.

(see commit at 0d8ac185cfeb6566a552ba97471d3b7dd23c8ac9)

> - Add a local storage state in RefinebioContext / exposed in hook with value of datasetAccessions that will default to an > empty dict. The keys will be accession codes and the values will be dataset ids.

A lookup table `datasetAccessions` local storage state has been added to `RefinebioContext` to store the accession code of processing one-off experiment and imported into `useDatasetManager` used by `usePollDatasetState`. 

(see commit at b204a5da3c5de25e55ddaa8f0cccc44ca1aa95c3)

> - add new method pollDatasetId
> - add new method pollDatasetAccession
>    - these functions should just set the datasetid and that is what the useEffect/polling logic should check

The new methods have been added and are called from `useEffect` of components that uses the `usePollDatasetState` hook to set `polledDatasetId` for polling.  

(see previous commit and at b204a5da3c5de25e55ddaa8f0cccc44ca1aa95c3)

> a separate cleanup function ie removeProcessingDataset that only takes dataset id and removes it from the list of processing datasets and the datasetAccessions object.

The clean up method has been added which removes `polledDatasetId` from:
- the `processingDatasets` array
- the `datasetAccessions` object

Please note that I've also included a line that resets the `polledDatasetId` to `null` to stop the running polling timer (initially added it to the cleanup block of `refreshProcessingDataset`, but moved it to this method).

This PR is ready for your review, thank you David!

davidsmejia commented on src/hooks/usePollDatasetStatus.js:
```diff
@@ -0,0 +1,81 @@
+  const [polledDatasetId, setPolledDatasetId] = useState(null)
+  const [polledDatasetState, setPolledDatasetState] = useState(false)
+  const timerRef = useRef(null)
+  const isProcessingDataset = polledDatasetState?.is_processing
```
```suggestion
  const [isProcessing, setIsProcessing] = useState(false)
```
If this is the only output we care about let's keep this a state and have `refreshProcessingDataset` handle setting this to true or false accordingly.

nozomione replied:
I've removed this local variable and added a new state `isProcessingDataset` (boolean), see at a5bf82650a3adc2e4925149acbcb0b3b4820ee21.

davidsmejia commented on src/hooks/usePollDatasetStatus.js:
```diff
@@ -0,0 +1,81 @@
+    getDataset
+  } = useDatasetManager()
+  const [polledDatasetId, setPolledDatasetId] = useState(null)
+  const [polledDatasetState, setPolledDatasetState] = useState(false)
```
```suggestion
```

We can remove this since we only really care about the status.

nozomione replied:
This `polledDatasetState` state is currently used to re-render `DatasetPageHeader`, please see the code block at aee530bbca025798b987282da8e92a45f2369c36.

davidsmejia commented on src/hooks/usePollDatasetStatus.js:
```diff
@@ -0,0 +1,81 @@
+
+  const refreshProcessingDataset = async () => {
+    const response = await getDataset(polledDatasetId)
+    // TEMP: until the fetchAsync is refactored
+    if (response?.ok !== false) {
+      setPolledDatasetState(response)
+    }
+
+    if (!response.is_processing) {
+      removeProcessingDataset()
+    }
```
```suggestion
    const newIsProcessing = response.is_processing !== false
    setIsProcessing(newIsProcessing)
    if (newIsProcessing) clearInterval(timerRef.current)
```

This should be enough here. All we want to know if is the dataset is still processing. The `useDatasetManager` will handle putting and removing the dataset from the processing list. We just need to update the current state after fetching and then clear the timer if we no longer need it. Arguably, this could be put in another useEffect but this is fine as is in my opinion.

nozomione replied:
Great, I used the separate `useEffect` as you suggested, see at bf3e091b238d595026c3e96e7a3ce86aa5b81f4d!

davidsmejia commented on src/components/shared/SearchCard/SearchCardAction/SearchCardAction.js:
```diff
@@ -15,16 +15,21 @@ export const SearchCardAction = ({
 }) => {
-  const { getProcessingExperiment } = useOneOffExperiment(accessionCode)
   const pageRendered = usePageRendered()
+  const { datasetAccessions, pollDatasetAccession } = usePollDatasetStatus()
```
```suggestion
  const { isProcessing, pollDatasetAccession } = usePollDatasetStatus()
```

nozomione replied:
Now using the new state `isProcessingDataset`, see at a5bf82650a3adc2e4925149acbcb0b3b4820ee21.

davidsmejia commented on src/hooks/usePollDatasetStatus.js:
```diff
@@ -0,0 +1,81 @@
+
+  // sets the mached dataset ID to polledDatasetId
+  const pollDatasetId = (datasetId) => {
+    if (processingDatasets.includes(datasetId)) setPolledDatasetId(datasetId)
```
If you also update `isProcessing` to true when the dataset id is in `processingDatasets`, then the exported `isProcessing` will trigger a render and display correctly for the time before the first interval is expired.

nozomione replied:
The new state  was added at a5bf82650a3adc2e4925149acbcb0b3b4820ee21.

nozomione replied:
See update at a5bf82650a3adc2e4925149acbcb0b3b4820ee21.

davidsmejia commented on src/hooks/usePollDatasetStatus.js:
```diff
@@ -0,0 +1,81 @@
+  // polls the latest state of the processing dataset per minute
+  // (the processing usually takes a few minutes)
+  useEffect(() => {
+    if (polledDatasetId) {
```
should eagerly clear the interval if `polledDatasetId` is true. This would prevent creating a new interval and losing the reference to the old one.

nozomione replied:
See update at 20607db98fa01e5661ccbbc8704d7c1f132774c5.

davidsmejia reviewed (commented):
Some comments around organization and improvement to the functionality of the hook.

The overall usage of the `usePollDatasetStatus` hook should be:

```javascript

const { isProcessing, pollDatasetAccession } = usePollDatasetStatus()

pollDatasetAccession(accession)

// logic using current state of `isProcessing`
```

nozomione commented on src/hooks/usePollDatasetStatus.js:
```diff
@@ -0,0 +1,81 @@
+
+    setProcessingDatasets((prev) => prev.filter((id) => id !== datasetId))
+    setPolledDatasetId(null) // removes polledDatasetId to stop the running timer
+  }
```
I've also added a new method  `getProcessingDatasetByAccession` which returns a processing dataset ID in the `datasetAccessions` look up table by an accession code which is used in `ProcessingDatasetPillModal` (see at a5bf82650a3adc2e4925149acbcb0b3b4820ee21) as chatted in 1:1.

nozomione commented on src/components/shared/SearchCard/SearchCardAction/SearchCardAction.js:
```diff
@@ -1,7 +1,7 @@
-import { memo } from 'react'
+import { memo, useEffect } from 'react'
```
Without `useEffect` and calling `pollDatasetAccession` will results in run-time error (`Too many re-renders`). 

Please see Vercel preview [here](https://refinebio-gr8s0h2ob-ccdl.vercel.app/search) (at 5804932e336bef1d6875eca092d248b44d26a1cc).

davidsmejia replied:
You can fix that by just adding a guard on `pollDatasetId`. Just return early if the passed in `datasetId` is already `polledDatasetId` since you can presume you already have a timer for that in place.

nozomione replied:
See update at 89f9daea7d91711677556287cf635924d38a667d.

nozomione commented on src/components/shared/SearchCard/SearchCardAction/SearchCardAction.js:
```diff
@@ -15,16 +15,21 @@ export const SearchCardAction = ({
+  useEffect(() => {
+    // sets a processing datasets for polling
+    pollDatasetAccession(accessionCode)
+  }, [accessionCode, datasetAccessions])
```
Please see https://github.com/AlexsLemonade/refinebio-web/pull/294#discussion_r1630376252.

nozomione commented:
I've updated the PR based on your feedback and ready for your review, thank you David!

davidsmejia reviewed (commented):
Responded to your question and added some suggestions for clarity.

nozomione commented on src/hooks/useDatasetManager.js:
```diff
@@ -100,10 +151,11 @@ export const useDatasetManager = () => {
+    const datasetIdToUse = isCurrentDatasetId ? id : await createDataset()
+    const response = await updateDataset(datasetIdToUse, params)
+    // adds this dataset ID to processingDatasets[] for polling
+    addToProcessingDatasets(datasetIdToUse, accessionCode)
```
See commit at e65aa12dff361a3983717deb1d28c03a937fe8cc.

nozomione replied:
See update at e65aa12dff361a3983717deb1d28c03a937fe8cc.

nozomione commented on src/hooks/useDatasetManager.js:
```diff
@@ -83,13 +124,23 @@ export const useDatasetManager = () => {
     }
 
+    // removes this dataset ID from processingDatasets[] if it exists
+    if (!response.is_processing) {
```
See update at 3bb2858f3df47a6c092cec26347139f5407984ee.

nozomione commented on src/hooks/useDatasetManager.js:
```diff
@@ -26,7 +30,44 @@ export const useDatasetManager = () => {
+
+        return temp
+      })
+    }
```
See updates at 4b4ceaf8a91c7f48b5c73ab13fcbb44c87aaa3cf and a41eb4ed1eb226a88f47006fa20f02c37f42df64.

davidsmejia commented on src/pages/dataset/[dataset_id].js:
Before we merge this, we need to clean this file up. I should be able to look at this and tell which dataset we are presenting here. As far as I can tell we should *only* care about the dataset from the query parameter. Please explain if I am missing something.

nozomione replied:
This page handles two types of datasets:

**Type 1:** `selectedDataset` - the dataset currently selected on the page

This state is used to store the dataset currently displayed on the page, which is the dataset ID obtained from the URL query parameter stored in `idFromQuery`.

<details>
  <summary>A dataset ID from URL</summary><br/>

A dataset ID `0a226474-fc40-4fe1-bc32-836ccb788b11` in URL:

![idFromQuery](https://github.com/AlexsLemonade/refinebio-web/assets/31800566/78845327-85c2-454e-ac86-57b3d151ee1d)

<hr/>
</details>

**Type 2:** `regeneratedDataset` - the dataset to be regenerated (see #287 for reference)

This state is used to store a modified version of the already processed dataset. Due to the immutability of the processed datasets, a new dataset ID will be generated if a user edits the download options before regenerating the processed dataset displayed on the page (`selectedDataset`). 

And we use this newly created dataset to re-render the modified dataset details of the processed dataset in the Download Files Summary section for that user.

<details>
  <summary><strong>Example</strong></summary><br/>

To replicate: 

**Step 1:** Visit https://refinebio-c1amq1arv-ccdl.vercel.app/dataset/0a94212a-04bf-4733-a3f6-0d9caed5a0bb

**Step 2:** Change the processed dataset's download options to see the UI update

Before updating the download options:

![BeforeUpdatingDownloadOptions](https://github.com/nozomione/ccdl/assets/31800566/97a57226-9b94-4e2e-9b39-11046dc4d051)

After updating the download options, a new dataset will be created behind the scene and stored in `regeneratedDataset`. We'll then swap the dataset currently displayed (`selectedDataset`) only in the Download Files Summary section to this new one to reflect the new changes in the UI accordingly:

![Screenshot 2024-06-13 at 5 12 44 PM](https://github.com/nozomione/ccdl/assets/31800566/d518f4e6-2443-4984-adad-27b8c6e61119)

<hr/>
</details>

**Other variables:**

 `isSameId` is used for the Move to Dataset button state. If the dataset ID currently displayed on the page is the same as My dataset ID (`datasetId`), then this button is disabled, otherwise is enabled. 

<details>
  <summary><strong>Example</strong></summary><br/>

To replicate: 

**Step 1:** Add an experiment to My Dataset

**Step 2:** Click the Share Dataset button to get a URL

**Step 3:** Open the URL in the browser

e.g.) If the dataset ID currently displayed is same as My dataset, we disable the Move to Dataset button:

![isSameId](https://github.com/nozomione/ccdl/assets/31800566/f1612268-18b4-4e3f-80b1-4abbcc0e9723)

<hr/>
</details>

`isUnprocessedDataset` is used to display the Download Dataset button. If the dataset currently displayed on the page is unprocessed (i.e. unprocessed shared dataset), we display this button in the UI, otherwise hide it.

<details>
  <summary><strong>Example</strong></summary><br/>

For unprocessed, the Download Dataset button is visible: 
https://refinebio-c1amq1arv-ccdl.vercel.app/dataset/7b501023-98f6-4f0f-b410-87fe4a6e5d19?ref=share

![ForUnprocessedDataset](https://github.com/nozomione/ccdl/assets/31800566/8a32e980-ff19-477b-90ed-d4da0c3460b0)

For processed dataset, it's hidden: 
https://refinebio-c1amq1arv-ccdl.vercel.app/dataset/0a226474-fc40-4fe1-bc32-836ccb788b11

![ForProcessedDataset](https://github.com/nozomione/ccdl/assets/31800566/58e2d26e-8b8a-4bef-94c5-dd62805c4edc)

<hr/>
</details>

 `isProcessed` is used to display the download options in the Download Files Summary section to regenerate the processed dataset.

<details>
   <summary><strong>Example</strong></summary><br/>

For processed dataset, the download options is visible: 
https://refinebio-c1amq1arv-ccdl.vercel.app/dataset/0a226474-fc40-4fe1-bc32-836ccb788b11

![downloadOptionVisible](https://github.com/nozomione/ccdl/assets/31800566/275dad6c-6199-47b3-997c-3c5933bbf193)

For unprocessed, it's hidden:
https://refinebio-c1amq1arv-ccdl.vercel.app/dataset/7b501023-98f6-4f0f-b410-87fe4a6e5d19?ref=share

![downloadOptionHidden](https://github.com/nozomione/ccdl/assets/31800566/17b0bffe-624e-49aa-9609-13d7f88d6277)

<hr/>
</details>
  
Please let me know if you have any feedback or questions, thank you David!

davidsmejia reviewed (approved):
Ok, that page should be reworked so that it isnt juggling variables like that. But we should handle that in a different issue and PR.

For the issue this PR is good for me
